### PR TITLE
[core] Fix Coverity auto cause copy in core sources

### PR DESCRIPTION
**Pull request by praasz**
### Details:
 - Apply changes to remove `auto_causes_copy` reported by Coverity in core sources.
 - Avoid copies also reduce bin-size.

### Tickets:
 - N/A

### CODE REVIEW

File: src/core/src/op/adaptive_avg_pool.cpp [hunk 1]
```diff
@@ -17,8 +17,8 @@ op::v8::AdaptiveAvgPool::AdaptiveAvgPool(const Output<Node>& data, const Output<
 void op::v8::AdaptiveAvgPool::validate_and_infer_types() {
     OV_OP_SCOPE(v8_AdaptiveAvgPool_validate_and_infer_types);
 
-    const auto output_shape = shape_infer(this, ov::util::get_node_input_partial_shapes(*this)).front();
-    set_output_type(0, get_input_element_type(0), output_shape);
+    const auto output_shapes = shape_infer(this, ov::util::get_node_input_partial_shapes(*this));
```
Comment (vurusovs) [inline src/core/src/op/adaptive_avg_pool.cpp:20]:
```suggestion
    const auto& output_shapes = shape_infer(this, ov::util::get_node_input_partial_shapes(*this));
```

File: src/core/src/op/adaptive_avg_pool.cpp [hunk 2]
```diff
@@ -17,8 +17,8 @@ op::v8::AdaptiveAvgPool::AdaptiveAvgPool(const Output<Node>& data, const Output<
...
+    set_output_type(0, get_input_element_type(0), output_shapes[0]);
```
Comment (vurusovs) [inline src/core/src/op/adaptive_avg_pool.cpp:21]: Why the change is better?

File: src/core/src/op/concat.cpp [hunk 3]
```diff
@@ -39,7 +39,8 @@ void Concat::validate_and_infer_types() {
         input_shapes.push_back(get_input_partial_shape(i));
     }
 
-    const auto output_shape = shape_infer(this, input_shapes).front();
+    const auto output_shapes = shape_infer(this, input_shapes);
```
Comment (vurusovs) [inline src/core/src/op/concat.cpp:42]:
```suggestion
    const auto& output_shapes = shape_infer(this, input_shapes);
```

File: src/core/src/op/space_to_batch.cpp [hunk 4]
```diff
@@ -55,8 +55,8 @@ void SpaceToBatch::validate_and_infer_types() {
                           "pads_end must be an integral number but got (",
                           pads_end_type,
                           ").");
-    const auto output_shape = shape_infer(this, ov::util::get_node_input_partial_shapes(*this)).front();
-    set_output_type(0, data_type, output_shape);
+    const auto& output_shapes = shape_infer(this, ov::util::get_node_input_partial_shapes(*this));
```
Comment (vurusovs) [inline src/core/src/op/space_to_batch.cpp:58]: In so many places with `shape_infer`, there are `auto& output_shapes` or `auto output_shapes` used. Was it intentional? May it be aligned?

File: src/core/src/op/adaptive_avg_pool.cpp (hunk 2 above)
Comment (praasz) [reply src/core/src/op/adaptive_avg_pool.cpp:21]: The `shape_infer` creates vector and `front` or `[0]` will make a copy this shape. Then this shape is passed as reference to `set_output_type` and copied again inside this setter.

New version will store vector and required shape is passed as reference and copied once inside `set_output_type`

File: src/core/src/op/space_to_batch.cpp (hunk 4 above)
Comment (praasz) [reply src/core/src/op/space_to_batch.cpp:58]: Yes will fix that. For store vector it should be fine to have `const auto output_shapes`.

File: src/core/src/op/depth_to_space.cpp [hunk 5]
```diff
@@ -42,8 +42,8 @@ std::shared_ptr<Node> DepthToSpace::clone_with_new_inputs(const OutputVector& ne
 void DepthToSpace::validate_and_infer_types() {
     OV_OP_SCOPE(v0_DepthToSpace_validate_and_infer_types);
 
-    const auto output_shape = shape_infer(this, ov::util::get_node_input_partial_shapes(*this)).front();
-    set_output_type(0, get_input_element_type(0), output_shape);
+    const auto output_shapes = shape_infer(this, ov::util::get_node_input_partial_shapes(*this));
```
Comment (rkazants) [inline src/core/src/op/depth_to_space.cpp:45]:
```suggestion
    const auto& output_shapes = shape_infer(this, ov::util::get_node_input_partial_shapes(*this));
```

File: src/core/src/op/gather_tree.cpp [hunk 6]
```diff
@@ -50,7 +50,7 @@ void op::v1::GatherTree::validate_and_infer_types() {
                           "Element type of inputs must be numeric. Got: ",
                           result_et);
 
-    const auto output_shape = shape_infer(this, ov::util::get_node_input_partial_shapes(*this)).front();
-    set_output_type(0, result_et, output_shape);
+    const auto output_shapes = shape_infer(this, ov::util::get_node_input_partial_shapes(*this));
```
Comment (rkazants) [inline src/core/src/op/gather_tree.cpp:53]:
```suggestion
    const auto& output_shapes = shape_infer(this, ov::util::get_node_input_partial_shapes(*this));
```

File: src/core/src/op/max_pool.cpp [hunk 7]
```diff
@@ -421,9 +421,9 @@ bool MaxPool::evaluate(TensorVector& outputs, const TensorVector& inputs) const
     const auto input_shapes = std::vector<PartialShape>{inputs[0].get_shape()};
     auto pads_begin = m_pads_begin;
     auto pads_end = m_pads_end;
-    const auto output_shape = shape_infer(this, input_shapes, pads_begin, pads_end).front();
+    const auto output_shapes = shape_infer(this, input_shapes, pads_begin, pads_end);
```
Comment (rkazants) [inline src/core/src/op/max_pool.cpp:424]:
```suggestion
    const auto& output_shapes = shape_infer(this, input_shapes, pads_begin, pads_end);
```

File: src/core/src/op/reverse.cpp [hunk 8]
```diff
@@ -63,8 +63,8 @@ void Reverse::validate_and_infer_types() {
                               "In 'index' mode the second input must contain integer values.");
     }
 
-    const auto output_shape = shape_infer(this, ov::util::get_node_input_partial_shapes(*this)).front();
-    set_output_type(0, get_input_element_type(0), output_shape);
+    const auto output_shapes = shape_infer(this, ov::util::get_node_input_partial_shapes(*this));
```
Comment (rkazants) [inline src/core/src/op/reverse.cpp:66]:
```suggestion
    const auto& output_shapes = shape_infer(this, ov::util::get_node_input_partial_shapes(*this));
```

File: src/core/src/op/roll.cpp [hunk 9]
```diff
@@ -28,8 +28,8 @@ void Roll::validate_and_infer_types() {
                           axes_et.is_dynamic() || axes_et == element::i32 || axes_et == element::i64,
                           "Axes must have int32 or int64 element type.");
 
-    const auto output_shape = shape_infer(this, ov::util::get_node_input_partial_shapes(*this)).front();
-    set_output_type(0, get_input_element_type(0), output_shape);
+    const auto output_shapes = shape_infer(this, ov::util::get_node_input_partial_shapes(*this));
```
Comment (rkazants) [inline src/core/src/op/roll.cpp:31]:
```suggestion
    const auto& output_shapes = shape_infer(this, ov::util::get_node_input_partial_shapes(*this));
```

File: src/core/src/op/shuffle_channels.cpp [hunk 10]
```diff
@@ -41,8 +41,8 @@ size_t ShuffleChannels::get_zero_based_axis() const {
 void ShuffleChannels::validate_and_infer_types() {
     OV_OP_SCOPE(v0_ShuffleChannels_validate_and_infer_types);
 
-    const auto output_shape = shape_infer(this, ov::util::get_node_input_partial_shapes(*this)).front();
-    set_output_type(0, get_input_element_type(0), output_shape);
+    const auto output_shapes = shape_infer(this, ov::util::get_node_input_partial_shapes(*this));
```
Comment (rkazants) [inline src/core/src/op/shuffle_channels.cpp:44]:
```suggestion
    const auto& output_shapes = shape_infer(this, ov::util::get_node_input_partial_shapes(*this));
```

File: src/core/src/preprocess/preprocess_impls.cpp [hunk 11]
```diff
@@ -324,7 +327,7 @@ void InputInfo::InputInfoImpl::dump(std::ostream& str,
 void OutputInfo::OutputInfoImpl::build(ov::ResultVector& results) {
     std::shared_ptr<opset8::Result> result;
     auto node = m_output_node;
-    auto start_out_node_names = node.get_tensor().get_names();
+    const auto start_out_node_names = node.get_tensor().get_names();
```
Comment (rkazants) [inline src/core/src/preprocess/preprocess_impls.cpp:330]:
```suggestion
    const auto& start_out_node_names = node.get_tensor().get_names();
```

Comment (praasz) [reply src/core/src/preprocess/preprocess_impls.cpp:330]: Here copy is intentional to restore names.

**Review by itikhono — APPROVED** [review]

**Comment by itikhono** [comment]
in general, LGTM

File: src/core/src/op/depth_to_space.cpp (hunk 5 above)
Comment (praasz) [reply src/core/src/op/depth_to_space.cpp:45]: All usage aligned to use `const auto` vector is created in this place the `&` is not required.

File: src/core/src/op/concat.cpp (hunk 3 above)
Comment (praasz) [reply src/core/src/op/concat.cpp:42]: Aligned to use `const auto`.

**Review by mitruska — APPROVED** [review]

File: src/core/src/op/slice.cpp [hunk 12]
```diff
@@ -123,9 +123,9 @@ bool Slice::evaluate(TensorVector& outputs, const TensorVector& inputs) const {
         shape_infer(this, ov::util::get_tensors_partial_shapes(inputs), make_tensor_accessor(inputs));
     outputs[0].set_shape(output_shapes.front().to_shape());
 
-    const auto starts = ov::get_tensor_data_as<int64_t>(inputs[1]);
-    const auto steps = ov::get_tensor_data_as<int64_t>(inputs[3]);
-    const auto axes = slice_no_axes(this) ? default_axes(starts.size()) : ov::get_tensor_data_as<int64_t>(inputs[4]);
+    const auto& starts = ov::get_tensor_data_as<int64_t>(inputs[1]);
+    const auto& steps = ov::get_tensor_data_as<int64_t>(inputs[3]);
+    const auto& axes = slice_no_axes(this) ? default_axes(starts.size()) : ov::get_tensor_data_as<int64_t>(inputs[4]);
```
Comment (t-jankowski) [inline src/core/src/op/slice.cpp:128]: Why is it needed? Doesn't it break RVO?

**Review by t-jankowski — APPROVED** [review]

File: src/core/src/op/slice.cpp (hunk 12 above)
Comment (praasz) [reply src/core/src/op/slice.cpp]: The `const auto&` will not break RVO. Is reported by Coverity (could be false positive). 
From binary size perspective there is no difference (chance code is same). Leavy as it can silence the Coverity issue and it looks like there is no impact on code.
